python/plugins/processing/algs/qgis/KNearestConcaveHull.py

```diff
@@ -0,0 +1,526 @@
+                                                            QgsProcessing.TypeVectorPolygon))
+
+    def processAlgorithm(self, parameters, context, feedback):
+        # get variables from dialog
+        source = self.parameterAsSource(parameters, self.INPUT, context)
+        if source is None:
+            raise QgsProcessingException(self.invalidSourceError(parameters, self.INPUT))
+
+        field_name = self.parameterAsString(parameters, self.FIELD, context)
+        kneighbors = self.parameterAsInt(parameters, self.KNEIGHBORS, context)
+        use_field = bool(field_name)
+
+        field_index = -1
+
+        fields = QgsFields()
+        fields.append(QgsField('id', QVariant.Int, '', 20))
+
+        # get properties of the field the grouping is based on
+        if use_field:
+            field_index = source.fields().lookupField(field_name)
+            if field_index >= 0:
+                fields.append(source.fields()[field_index])
+
+        # initialize writer
+        (sink, dest_id) = self.parameterAsSink(parameters, self.OUTPUT, context,
+                                               fields, QgsWkbTypes.Polygon, source.sourceCrs())
+
+        current = 0
+        fid = 0
+
+        if field_index >= 0:
+            # get unique values of field denoted by index as filter conditions
+            unique_values = source.uniqueValues(field_index)
+            total = 100.0 / float(source.featureCount() * len(unique_values))
+
+            for unique in unique_values:
+                points = []
+                filter = QgsExpression.createFieldEqualityExpression(field_name, unique)
+                request = QgsFeatureRequest().setFilterExpression(filter)
+                request.setSubsetOfAttributes([])
+                features = source.getFeatures(request)
+                for in_feature in features:
+                    points.extend(extract_points(in_feature.geometry()))
+                    current += 1
+                    feedback.setProgress(int(current * total))
+
+                # A minimum of 3 points is necessary to proceed
+                if len(points) >= 3:
+                    out_feature = QgsFeature()
+                    try:
+                        the_hull = concave_hull(points, kneighbors)
+                        if the_hull:
+                            vertex = [QgsPointXY(point[0], point[1]) for point in the_hull]
+                            poly = QgsGeometry().fromPolygonXY([vertex])
+
+                            out_feature.setGeometry(poly)
+                            out_feature.setAttributes([fid, unique])
+                            sink.addFeature(out_feature, QgsFeatureSink.FastInsert)
+                    except:
+                        feedback.reportError('Exception while computing concave hull.')
+                        raise QgsProcessingException('Exception while computing concave hull.')
+                fid += 1
+
+        else:
+            points = []
```

Comment:
Hi @rudivs, I'm not completly inside the logic of the alg, but is this part exactly the same as above executed "field_index >= 0"? the only change is the request construction. is it possible to avoid this redundancy?

Reply:
@luipir this second section echoes the same basic logic, but the first section has extra steps for filtering features by the grouping field and then adding that as an attribute to the output polygons (which may be multiple). It might be possible to extract this as a function which has slightly different behaviour based on whether there's a grouping field set or not, but I think that might make it harder to parse.